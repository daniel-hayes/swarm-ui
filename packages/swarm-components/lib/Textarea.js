"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var getTextareaStatus = function getTextareaStatus(props) {
  return props.disabled ? 'disabled' : props.error ? 'error' : 'default';
};

var getCharacterCount = function getCharacterCount() {
  var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  return value.length;
};

var Textarea = function Textarea(props) {
  console.log('textarea char count', getCharacterCount(props.value));
  return React.createElement("textarea", (0, _extends2["default"])({
    "data-swarm-textarea": getTextareaStatus(props)
  }, props));
};

var _default = Textarea;
exports["default"] = _default;
Textarea.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Textarea",
  "props": {
    "disabled": {
      "required": false,
      "flowType": {
        "name": "boolean"
      },
      "description": "Whether the textarea should be interactive."
    },
    "error": {
      "required": false,
      "flowType": {
        "name": "boolean"
      },
      "description": "Whether the textarea renders an error state."
    },
    "value": {
      "required": false,
      "flowType": {
        "name": "string"
      },
      "description": "Value for textarea."
    }
  }
};