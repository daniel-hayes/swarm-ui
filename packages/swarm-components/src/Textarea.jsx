// @flow
import * as React from 'react';

type Props = {
    disabeld?: boolean,
    error?: boolean,
    value?: string
}

const getTextareaStatus = (props: Props) => {
    let status = 'default';
    if (props.error) status = 'error';
    if (props.disabeld) status = 'disabled';
    return status;
}

const getCharacterCount = (value: string = '') => value.length;

export const Textarea = (props: Props) => {

    console.log('textarea char count', getCharacterCount(props.value))

    return <textarea data-swarm-textarea={getTextareaStatus(props)} {...props} />
}

export default Textarea;