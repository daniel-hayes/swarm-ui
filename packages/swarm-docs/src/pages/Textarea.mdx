import { Textarea } from '@meetup/swarm-components'
import { Icon } from '@meetup/swarm-components'
import Example from '../components/textareaExample';
import PropsTable from '../components/propstable';

# Textarea

```.jsx
<Textarea placeholder='textarea component for v2' />
```

```.jsx
<Textarea value='textarea component for v2' />
```

## Example
<Example />


### Props
---
<br/>
<PropsTable propMetaData={Textarea.__docgenInfo} />
