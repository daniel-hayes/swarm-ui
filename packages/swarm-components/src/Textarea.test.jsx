import React from 'react';
import { shallow, mount } from 'enzyme';
import Textarea, {
	getTextareaStatus,
} from './Textarea';
import auto from 'autosize';

// mock `auto` from autosize library
jest.mock('autosize', () => {
	let autosize = jest.fn();
	autosize.update = jest.fn();
	autosize.destory = jest.fn();
	return autosize;
});

describe('Textarea', () => {
    const mockOnChange = jest.fn();

	const testCases = [
		['Default (empty)', <Textarea key="0" value="" />],
		['Default (value)', <Textarea key="1" value="foo" />],
		['Disabled', <Textarea key="2" value="" disabled />],
		['Error', <Textarea key="3" value="foo" error />],
		['maxLength', <Textarea key="4" value="foo" maxLength="20" />],
		['autosize', <Textarea key="5" value="foo" autosize />],
		['onChange', <Textarea key="6" value="" onChange={mockOnChange} />],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
    });

    describe('autosize', ()=> {
        it('calls autosize on mount', () => {
			auto.mockClear();
            mount(<Textarea value="something" onChange={mockOnChange} autosize />);
            expect(auto).toHaveBeenCalled();
		});

		it('calls autosize.update on componentDidUpdate', () => {
			auto.mockClear();
            const textarea = mount(<Textarea value="something" onChange={mockOnChange} autosize />);
			textarea.instance().componentDidUpdate({value: "prev props values"});
			expect(auto.update).toHaveBeenCalled();
		});
    });
});

describe('getTextareaStatus', () => {
	it('returns disabled when props contain disabled=true', () => {
		expect(getTextareaStatus({disabled: true})).toEqual('disabled');
		expect(getTextareaStatus({disabled: true, error: true})).toEqual('disabled');
	});

	it('returns error when props contain error=true', () => {
		expect(getTextareaStatus({error: true})).toEqual('error');
	});

	it('returns default when props do not contain disabled or error', () => {
		expect(getTextareaStatus({})).toEqual('default');
	});
});
