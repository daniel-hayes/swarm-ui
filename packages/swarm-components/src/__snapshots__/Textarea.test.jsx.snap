// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Snapshot: Default (empty) 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="default"
    value=""
  />
</div>
`;

exports[`Textarea Snapshot: Default (value) 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="default"
    value="foo"
  />
</div>
`;

exports[`Textarea Snapshot: Disabled 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="disabled"
    disabled={true}
    value=""
  />
</div>
`;

exports[`Textarea Snapshot: Error 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="error"
    error={true}
    value="foo"
  />
</div>
`;

exports[`Textarea Snapshot: autosize 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="default"
    value="foo"
  />
</div>
`;

exports[`Textarea Snapshot: maxLength 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="default"
    value="foo"
  />
  <CharCount>
    NaN
  </CharCount>
</div>
`;

exports[`Textarea Snapshot: onChange 1`] = `
<div
  data-swarm-textarea-wrapper={true}
>
  <textarea
    data-swarm-textarea="default"
    onChange={[MockFunction]}
    value=""
  />
</div>
`;
